src/transformers/models/xglm/modeling_xglm.py

```diff
@@ -935,16 +857,26 @@ def forward(
     def prepare_inputs_for_generation(
         self, input_ids, past_key_values=None, attention_mask=None, use_cache=None, **kwargs
     ):
-        # if model is used as a decoder in encoder-decoder model, the decoder attention mask is created on the fly
-        if attention_mask is None:
-            attention_mask = input_ids.new_ones(input_ids.shape)
+        position_ids = kwargs.get("position_ids", None)
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                position_ids = position_ids[:, -1].unsqueeze(-1)
+        else:
+            position_ids = None
```

Comment:
I'm a bit confused 😅  Why do we set `position_ids` as `None` if they're not `None` when passed in?

Reply:
More than a valid question, this is a copy-paste from other models that (IMO) should be changed!

`position_ids` should be computed from `attention_mask`, so the `else` path invalidates them when `attention_mask is None`. However, that doesn't explain the `and position_ids is None` in the `if` 🙃 

A better change would be to move this `position_ids` computation to the modeling code itself, for all models. However, I'm not entirely sure whether researchers would miss being able to control `position_ids`.